### ISSUE TITLE
[Drain Mode] Functional Issues in Drain mode feature

### ISSUE THREAD

**Issue by rajiesh** [gocd/gocd#5633]
##### Issue Type

- Bug Report

##### Summary

Placeholder issue for reporting any functional issues with the drain mode feature. Please add details of the issue to the comments and add a point to below with link to the relevant comment

  - [x] Drain mode SPA should not show pipelines which are deleted from cruise config xml(https://github.com/gocd/gocd/issues/5633#issuecomment-450142764)
  - [x] When under server drain mode, need a consistent way to show drain mode message. Currently each page shows message `Server in drain mode` differently (https://github.com/gocd/gocd/issues/5633#issuecomment-450143553). This will be worked on as separate issue as all pages need to have one way to show system status like `backup in progress`, `maintenance mode`
 - [x] Jobs which are scheduled but not assigned to agent should not be included in the running stages of drain mode

**Comment by rajiesh** [comment]
Drain mode SPA should not show pipelines which are deleted from cruise config xml

Steps to reproduce:

1. Create a pipeline and start it to build
2. Delete the pipeline while its building
3. Enable server drain mode
4. Drain mode will never be complete, will be waiting on the deleted pipeline

**Comment by rajiesh** [comment]
When server under drain mode visit plugins page and try making any plugin settings change. Save will fail, however the page will not show any error messages. 
On Config repositories page if user create config repo, when server under drain mode, error is on the config repositories main page and not on the modal
On any of the admin pages like `User Summary`, `Config XML` etc the error is shown differently

Need to think about a consistent way to show server drain mode warning when user perform any Create/Update operations

**Comment by GaneshSPatil** [comment]
@rajiesh --
> Issue1: Drain mode SPA should not show pipelines which are deleted from cruise config xml

This is fixed as part of PR: #5641, As the deleted pipelines will not be cached as part of the dashboard. It won't be shown on the Drain mode SPA. Though, it opens up another issue for us: https://github.com/gocd/gocd/issues/5643


> Issue2: When under server drain mode, need a consistent way to show drain mode message

Currently, there are two ways to render a drain-mode in-progress error on a POST call. 
On all the new SPAs, if AJAX post call fails, it renders error flash message on the page.
Whereas, on old rails pages or vm pages, if page post fails, it does a page redirect to `drain-mode in-progress` page.
Also, on the newer SPA, its each page's responsibility to render any error occured during a POST call.

> Issue3: Jobs which are scheduled but not assigned to agent should not be included in the running stages of drain mode

IMO, it should be shown on the page, as its is a **RUNNING** job. Even though it wont affect the server state by any means, but the state is still running :)

**Comment by rajiesh** [comment]
> Issue3: Jobs which are scheduled but not assigned to agent should not be included in the running stages of drain mode

@GaneshSPatil thinking again on this issue, am still doubtful about including the scheduled jobs under running stages. I agree with your point that `scheduled/assigned/building/completed` are all internal states of pipeline and user need not bother of it. However when seeing from Drain mode feature perspective as a super admin user who is trying to set the server in drain mode, the expectation is that the drain mode shows as completed and then can proceed with restart/upgrade. Its highly likely in large setups during the time admin user enables drain mode there are pipelines in scheduled state. Which will never complete and drain mode will never show as successfully completed.

> IMO, it should be shown on the page, as its is a RUNNING job. Even though it wont affect the server state by any means, but the state is still running :)

Also as you have mentioned above having pipeline which are scheduled but not assigned to agent does not affect server state, which then means those pipelines are not of concern to keep drain mode in non-completed state ie., Server is good to be taken down for maintenance even though there are pipelines in scheduled but not assigned state. So why even include pipeline state which are not a concern for drain mode?

cc: @arvindsv @maheshp
Including Arvind and Mahesh in case they have any opinion

**Comment by GaneshSPatil** [comment]
Fixed as part of #5629 #5641 

Verified on GoCD Version: 19.1.0 (8315-49ddcb8a0a7c3c8fe85025b8ca2b443a5dfdbd62).
